### ISSUE TITLE
Add dropout to tfjs-core.

### ISSUE THREAD

**Issue by nsthorat** [tensorflow/tfjs#117]
_From @akshay951228 on February 6, 2018 5:27_

I'm importing tensor model weight in DLJs, in my model I have dropout but dropout is not implemented yet.So my question is when it will be implemented

_Copied from original issue: tensorflow/tfjs-core#659_

**Comment by nsthorat** [comment]
_From @dsmilkov on February 8, 2018 23:56_

Dropout is a great example for an external contribution. Happy to provide review!

**Comment by nsthorat** [comment]
@caisq is this something that would be easy to push into deeplearnjs?

**Comment by nsthorat** [comment]
_From @caisq on February 9, 2018 2:37_

@nsthorat Feel free to assign this FR to me.

**Comment by nsthorat** [comment]
_From @dsmilkov on February 9, 2018 2:57_

Awesome. Trying to assign you but looks like we need to give you write access - will do now

**Comment by nsthorat** [comment]
_From @abrad1212 on February 23, 2018 3:40_

Couldn't you implement Dropout with everything in the current API?

Wouldn't it be similar to [Tensorflow's implementation](https://github.com/tensorflow/tensorflow/blob/r1.5/tensorflow/python/ops/nn_ops.py#L2135-L2196)?

**Comment by nsthorat** [comment]
_From @caisq on February 23, 2018 3:42_

@abrad212 thanks for the ping. I have this on my TODO list

**Comment by nsthorat** [comment]
_From @abrad1212 on February 23, 2018 3:44_

@caisq No problem. Anything to help 😄

**Comment by nsthorat** [comment]
_From @abrad1212 on February 23, 2018 17:40_

@caisq I'm traveling today, and I think I have time over this weekend to try to implement DropOut. Hopefully, the implementation will be similar to the link I posted above, and translate easy. I'll let you know if I make any progress.

**Comment by nsthorat** [comment]
_From @acostin1 on February 25, 2018 1:46_

Would love to test it, I think I need to use dropout layers because my model is over-fitting too fast. Also, let's make sure one can use it in the model-builder demo :)

**Comment by nsthorat** [comment]
_From @kwhitley on March 14, 2018 20:51_

Agreed re. all... definitely need dropout implementation stat.  The basic models without dropout, regularization, noise, etc overfit like a champ.  

😞

**Comment by nsthorat** [comment]
Coming soon, I promise!

**Comment by nsthorat** [comment]
_From @caisq on April 1, 2018 15:21_

FYI, with the release of TensorFlow.js, now one can do dropout at the Layers level with `tf.layers.dropout`: https://js.tensorflow.org/api/0.6.1/#layers.dropout

As such, I'm removing the `P1` tag from this bug. Dropout can later be pushed down to tfjs-core, so that those who don't use `tfjs-layers` can also use it.

**Comment by VariableVasasMT** [comment]
picking up this issue

**Comment by caisq** [comment]
@VariableVasasMT Thanks for volunteering. Please note that the code for dropout exists in the tf.layers.Dropout layer in the tfjs-layers repo. I think the code can be migrated to tfjs-core with relatively minor changes.

**Comment by josephch405** [comment]
When you mean relatively minor changes, does this still involve adding a specific dropout implementation in both backends? Seems like a much more involved process than what they had in the layers version. Also still getting the hang of contributing to tfjs, so I'm just making sure I'm not misunderstanding anything about the codebase.

**Comment by euler16** [comment]
@VariableVasasMT are you still working on this issue?

**Comment by VariableVasasMT** [comment]
@euler16 yes, I am still working on it, as suggested by @caisq, I was reading the code to understand better but I would be raising a pull request soon.

**Comment by nsthorat** [comment]
You want to remove all the dependencies of tf.layers -- it shouldn't just be a code transplant, rather a reimplementation with core ops. Does that make sense?

**Comment by VariableVasasMT** [comment]
> You want to remove all the dependencies of tf.layers -- it shouldn't just be a code transplant, rather a reimplementation with core ops. Does that make sense?

That's fair, I am worried about two codes which do the same thing and are very similar would exist parallelly. I am pro for reimplementation with core ops, just wanted to confirm once.

**Comment by nsthorat** [comment]
Once the core op is implemented we can call into it from layers :)

**Comment by VariableVasasMT** [comment]
@nsthorat please review the updated the pull request.

**Comment by syt123450** [comment]
Hi @nsthorat @caisq @dsmilkov , I sent a followup PR [tfjs-core#1782](https://github.com/tensorflow/tfjs-core/pull/1782) to make dropout op support non-default `noiseShape`, could you please take a look into it? Thanks!
